js_modules/dagit/packages/ui/src/components/MiddleTruncate.tsx

```diff
@@ -9,14 +8,87 @@ interface Props {
   text: string;
 }
 
+const FILL = 'M'.repeat(200);
+
 const getFont = (element: HTMLDivElement) => getComputedStyle(element).font;
 
+/**
+ * A component that performs middle truncation on a given string, based on the evaluated width
+ * of a container div.
+ *
+ * The component will briefly render a "fill" string to determine the maximum available width,
+ * then use that measurement and the computed font style to determine the longest middle-truncated
+ * string that can fit within that width.
+ *
+ * When the DOM element resizes, the measurement and calculation steps will occur again.
+ */
 export const MiddleTruncate: React.FC<Props> = (props) => {
   const {text} = props;
+
+  // Track the font style and target maximum width. `null` means no measurement has
+  // taken place yet.
   const [targetStyle, setTargetStyle] = React.useState<TargetStyle | null>(null);
-  const ref = React.useRef<HTMLDivElement>(null);
 
-  const truncated = useFixedSpan(text, targetStyle);
+  // Determine the width available to the text container. This renders a long string into
+  // the container to allow quickly measuring the maximum width available, which will then
+  // be used to determine the truncated string.
+  const [determineAvailableWidth, setDetermineAvailableWidth] = React.useState(true);
+
+  const container = React.useRef<HTMLDivElement>(null);
+
+  // Track whether we have just evaluated the avilable width of the container. If so,
+  // we want to ignore any resize observations temporarily, to avoid an observer->resize loop.
+  const justDetermined = React.useRef(false);
+
+  // Given the target font style and allotted width, calculate the largest possible middle-
+  // truncated string. This hook returns null if there is no style available yet.
+  const truncated = useMiddleTruncatedText(text, targetStyle);
+
+  // When we are rendering the "fill" string to determine the available width, use a layout
+  // effect to perform the measurement. Get the computed font style as well -- both of these
+  // will be used to draw the text string onto a canvas context to find the appropriate
+  // truncated string.
+  React.useLayoutEffect(() => {
+    if (container.current && determineAvailableWidth) {
```

Comment:
I think if you want, you could actually do this without the temporary state + re-render by directly assigning long content here, since the clientRect functions are synchronous:

```
      container.current.innerText = 'M'.repeat(200);
      setTargetStyle({
        width: container.current.getBoundingClientRect().width,
        font: getFont(container.current),
      });
      container.current.innerText = ''; // clear it, just to be sure nobody ever sees MMMMMM
      // innerText is restored by the re-render after setState
```

Reply:
Ah, that's a nice idea. I'll try it.